livekit-android-sdk/src/main/java/io/livekit/android/room/track/RemoteTrackPublication.kt

```diff
@@ -88,9 +92,16 @@ class RemoteTrackPublication(
         }
 
     /**
-     * subscribe or unsubscribe from this track
+     * Subscribe or unsubscribe from this track
+     *
+     * If [subscriptionAllowed] is false, subscription will fail.
      */
     fun setSubscribed(subscribed: Boolean) {
+        if (subscribed && !subscriptionAllowed) {
```

Comment:
Is this preventing sending an `UpdateSubscriptions` message? Sorry, I maybe mistaking code here. But, sending `UpdateSubscriptions` should not be prevented. Server will send the `SubscriptionPermissionUpdate` when there are no permissions. Note that permissions can be revoked when a track is actively subscribed. So, the server could send that update mid stream too. So, clients should always send subscription message and listen for the update from the server.

I think this code is not preventing sending that message, but just wanted to elaborate.

Reply:
Got it. Maybe, the log message can be something like `already subscribed, but currently does not have permissions`?

Reply:
I think we should still let the client update the server, because it's indicating a *desired* state.

It does make the API a bit inconsistent, which I don't love:

* .subscribed indicates if it's desired to be subscribed and successfully subscribed
* .setSubscribed updates the desired state

Any alternatives?

Reply:
Discussed with @boks1971, the proposal is to turn `.subscribed` into an enum describing the current subscription status, with "subscribed, not allowed" and "subscribed" as two distinct states.

Thinking about the `setSubscribed` part, I think it might be more descriptive of user intent to have separate `subscribe()` and `unsubscribe()` functions, which would help clear up the inconsistency you note.

Reply:
that sounds good. though I'm wary about breaking current users. Especially on JS. we could perhaps deprecate .subscribed, but still provide it as a way of returning `subscribeState == Subscribed` ?